Add new activation script and/or run dependency?

Continuing the discussion from https://github.com/AnacondaRecipes/cudatoolkit-feedstock/pull/5#issuecomment-603454124 and https://github.com/cupy/cupy/issues/3222#issuecomment-603449808:
1. It'd be much better if `nvcc` can exist in a conda env where CuPy is installed
2. If not, we should set `CUDA_PATH` for CuPy to look up where CUDA is in the conda env.

I see, thanks for confirming, John & Keith!

I am asking because this CuPy test is failing with CF's `nvcc_linux-64`, `gcc_linux-64`, `gxx_linux-64`, `cupy=7.3.0`, `cudatoolkit=10.0`:
```bash
$ pytest cupy_tests/core_tests/test_raw.py

(...omit output...)

>                   raise cex
E                   cupy.cuda.compiler.CompileException: `nvcc` command returns non-zero exit status. 
E                   command: ['/home/leofang/miniconda3/envs/CF_cupy_test2/bin/nvcc', '-gencode=arch=compute_70,code=sm_70', '--ptx', '-I/home/leofang/miniconda3/envs/CF_cupy_test2/lib/python3.7/site-packages/cupy/core/include', '-I /home/leofang/miniconda3/envs/CF_cupy_test2/lib/python3.7/site-packages/cupy/core/include/cupy/_cuda/cuda-10.0', '-I /usr/local/cuda/include', '-ftz=true', '/tmp/tmp6r4ykd15/preprocess.cu']
E                   return-code: 1
E                   stdout/stderr: 
E                   cc1plus: fatal error: cuda_runtime.h: No such file or directory
E                   compilation terminated.

../../miniconda3/envs/CF_cupy_test2/lib/python3.7/site-packages/cupy/cuda/compiler.py:198: CompileException
======================================================================= short test summary info =======================================================================
FAILED cupy_tests/core_tests/test_raw.py::TestRaw_param_1_{backend='nvcc'}::test_basic - cupy.cuda.compiler.CompileException: `nvcc` command returns non-zero exit s...
FAILED cupy_tests/core_tests/test_raw.py::TestRaw_param_1_{backend='nvcc'}::test_compiler_flag - cupy.cuda.compiler.CompileException: `nvcc` command returns non-zer...
FAILED cupy_tests/core_tests/test_raw.py::TestRaw_param_1_{backend='nvcc'}::test_cuDoubleComplex - cupy.cuda.compiler.CompileException: `nvcc` command returns non-z...
FAILED cupy_tests/core_tests/test_raw.py::TestRaw_param_1_{backend='nvcc'}::test_cuFloatComplex - cupy.cuda.compiler.CompileException: `nvcc` command returns non-ze...
FAILED cupy_tests/core_tests/test_raw.py::TestRaw_param_1_{backend='nvcc'}::test_dynamical_parallelism - cupy.cuda.compiler.CompileException: `nvcc` command returns...
FAILED cupy_tests/core_tests/test_raw.py::TestRaw_param_1_{backend='nvcc'}::test_dynamical_parallelism_compile_failure - cupy.cuda.compiler.CompileException: `nvcc`...
FAILED cupy_tests/core_tests/test_raw.py::TestRaw_param_1_{backend='nvcc'}::test_get_function_failure - cupy.cuda.compiler.CompileException: `nvcc` command returns ...
FAILED cupy_tests/core_tests/test_raw.py::TestRaw_param_1_{backend='nvcc'}::test_invalid_compiler_flag - cupy.cuda.compiler.CompileException: `nvcc` command returns...
FAILED cupy_tests/core_tests/test_raw.py::TestRaw_param_1_{backend='nvcc'}::test_kernel_attributes - cupy.cuda.compiler.CompileException: `nvcc` command returns non...
FAILED cupy_tests/core_tests/test_raw.py::TestRaw_param_1_{backend='nvcc'}::test_module - cupy.cuda.compiler.CompileException: `nvcc` command returns non-zero exit ...
FAILED cupy_tests/core_tests/test_raw.py::TestRaw_param_1_{backend='nvcc'}::test_module_both_code_and_path - cupy.cuda.compiler.CompileException: `nvcc` command ret...
FAILED cupy_tests/core_tests/test_raw.py::TestRaw_param_1_{backend='nvcc'}::test_module_load_failure - cupy.cuda.compiler.CompileException: `nvcc` command returns n...
FAILED cupy_tests/core_tests/test_raw.py::TestRaw_param_1_{backend='nvcc'}::test_module_neither_code_nor_path - cupy.cuda.compiler.CompileException: `nvcc` command ...
=================================================================== 13 failed, 13 passed in 11.03s ====================================================================
```

What these tests do is to use `RawModule` and `RawKernel` to compile some CUDA codes with both NVRTC and nvcc backends. The former works just fine, but CF's `nvcc` fails all tests due to the absence of CUDA headers. (To reproduce this, you need to mask your `CUDA_PATH` and `/usr/local/cuda` so that `get_cuda_path()` returns `None`.)

@jakirkham It seems patching `get_cuda_path()` as you suggested in https://github.com/cupy/cupy/issues/3222#issuecomment-603436868 would not help revolve this, because the CUDA headers are simply unavailable. We likely need to implement a detection scheme to tell if we're using CF `cupy`, and simply ban the usage of the `nvcc` backend in `RawModule` and `RawKernel`.  

However, your `sys.prefix` solution as well as the two of mine are still valid proposals for runtime discovery, which is needed to tell NVRTC where cudadevrt is, for example.

--- a/.scripts/build_steps.sh
+++ b/.scripts/build_steps.sh
@@ -19,7 +19,8 @@ conda-build:
 
 CONDARC
 
-conda install --yes --quiet conda-forge-ci-setup=2 conda-build -c conda-forge
+conda install --yes --quiet conda-forge-ci-setup=3 conda-build pip -c conda-forge
+
 
 # set up the condarc
 setup_conda_rc "${FEEDSTOCK_ROOT}" "${RECIPE_ROOT}" "${CONFIG_FILE}"
@@ -33,7 +34,7 @@ conda build "${RECIPE_ROOT}" -m "${CI_SUPPORT}/${CONFIG}.yaml" \
     --clobber-file "${CI_SUPPORT}/clobber_${CONFIG}.yaml"
 
 if [[ "${UPLOAD_PACKAGES}" != "False" ]]; then
-    upload_package "${FEEDSTOCK_ROOT}" "${RECIPE_ROOT}" "${CONFIG_FILE}"
+    upload_package  "${FEEDSTOCK_ROOT}" "${RECIPE_ROOT}" "${CONFIG_FILE}"
 fi
 
 touch "${FEEDSTOCK_ROOT}/build_artifacts/conda-forge-build-done-${CONFIG}"

--- a/.scripts/run_docker_build.sh
+++ b/.scripts/run_docker_build.sh
@@ -52,8 +52,10 @@ mkdir -p "$ARTIFACTS"
 DONE_CANARY="$ARTIFACTS/conda-forge-build-done-${CONFIG}"
 rm -f "$DONE_CANARY"
 
+# Allow people to specify extra default arguments to `docker run` (e.g. `--rm`)
+DOCKER_RUN_ARGS="${CONDA_FORGE_DOCKER_RUN_ARGS}"
 if [ -z "${CI}" ]; then
-    DOCKER_RUN_ARGS="-it "
+    DOCKER_RUN_ARGS="-it ${DOCKER_RUN_ARGS}"
 fi
 
 export UPLOAD_PACKAGES="${UPLOAD_PACKAGES:-True}"

--- a/recipe/activate.sh
+++ b/recipe/activate.sh
@@ -1,0 +1,3 @@
+#!/bin/bash
+
+export CUDA_PATH=$CONDA_PREFIX

--- a/recipe/activate.sh
+++ b/recipe/activate.sh
@@ -1,3 +1,8 @@
 #!/bin/bash
 
+
+if [[ -z "$CUDA_PATH"]]; then
+    export CONDA_CUPY_CUDA_PATH=$CUDA_PATH
+fi
+
 export CUDA_PATH=$CONDA_PREFIX

--- a/recipe/activate.sh
+++ b/recipe/activate.sh
@@ -1,7 +1,7 @@
 #!/bin/bash
 
 
-if [[ -z "$CUDA_PATH"]]; then
+if [[ -z $CUDA_PATH ]]; then
     export CONDA_CUPY_CUDA_PATH=$CUDA_PATH
 fi
 

--- a/recipe/deactivate.sh
+++ b/recipe/deactivate.sh
@@ -1,3 +1,9 @@
 #!/bin/bash
 
-unset CUDA_PATH
+
+if [[ -z "$CONDA_CUPY_CUDA_PATH"]]; then
+    export CUDA_PATH=$CONDA_CUPY_CUDA_PATH
+    unset CONDA_CUPY_CUDA_PATH
+else
+    unset CUDA_PATH
+fi

--- a/recipe/deactivate.sh
+++ b/recipe/deactivate.sh
@@ -1,0 +1,3 @@
+#!/bin/bash
+
+unset CUDA_PATH

--- a/recipe/deactivate.sh
+++ b/recipe/deactivate.sh
@@ -1,7 +1,7 @@
 #!/bin/bash
 
 
-if [[ -z "$CONDA_CUPY_CUDA_PATH"]]; then
+if [[ -z $CONDA_CUPY_CUDA_PATH ]]; then
     export CUDA_PATH=$CONDA_CUPY_CUDA_PATH
     unset CONDA_CUPY_CUDA_PATH
 else

--- a/recipe/meta.yaml
+++ b/recipe/meta.yaml
@@ -21,7 +21,7 @@ source:
   #  folder: cub_v1.8.0  # so the source code is in cupy/cub_v1.8.0
 
 build:
-  number: 0
+  number: 1
   # temporarily disable PyPy build (conda-forge/cupy-feedstock#43)
   skip: True  # [py27 or (not linux64) or cuda_compiler_version == "None" or python_impl == 'pypy']
   script:
@@ -32,6 +32,12 @@ build:
     - export CXXFLAGS="${CXXFLAGS} -I/usr/include"
     #- export CUB_PATH=${PWD}/cub_v1.8.0  # TODO(leofang): remove this
     - "{{ PYTHON }} -m pip install . -vv"
+
+    # copy activate/deactivate scripts
+    - mkdir -p "${PREFIX}/etc/conda/activate.d"
+    - cp "${RECIPE_DIR}/activate.sh" "${PREFIX}/etc/conda/activate.d/cupy_activate.sh"
+    - mkdir -p "${PREFIX}/etc/conda/deactivate.d"
+    - cp "${RECIPE_DIR}/deactivate.sh" "${PREFIX}/etc/conda/deactivate.d/cupy_deactivate.sh"
   missing_dso_whitelist:
     - "*/libcuda.*"
 

--- a/recipe/run_test.py
+++ b/recipe/run_test.py
@@ -5,6 +5,7 @@ os.environ["CUPY_TEST_GPU_LIMIT"] = "1"
 # Check CUDA_PATH is set
 cuda_path = os.environ.get('CUDA_PATH')
 assert cuda_path is not None
+print("CUDA_PATH:", cuda_path)
 
 # Check CUDA libraries are available
 assert os.path.isfile(os.path.join(cuda_path, 'lib/libcudart.so'))

--- a/recipe/run_test.py
+++ b/recipe/run_test.py
@@ -6,6 +6,9 @@ os.environ["CUPY_TEST_GPU_LIMIT"] = "1"
 cuda_path = os.environ.get('CUDA_PATH')
 assert cuda_path is not None
 
+# Check CUDA libraries are available
+assert os.path.isfile(os.path.join(cuda_path, 'lib/libcudart.so'))
+
 # Check for CuPy (without importing)
 import pkgutil
 pkgutil.find_loader("cupy")

--- a/recipe/run_test.py
+++ b/recipe/run_test.py
@@ -2,6 +2,10 @@
 import os
 os.environ["CUPY_TEST_GPU_LIMIT"] = "1"
 
+# Check CUDA_PATH is set
+cuda_path = os.environ.get('CUDA_PATH')
+assert cuda_path is not None
+
 # Check for CuPy (without importing)
 import pkgutil
 pkgutil.find_loader("cupy")

